datashader/core.py

```diff
@@ -196,11 +199,113 @@ def line(self, source, x, y, agg=None):
+        agg : Reduction, optional
+            Reduction to compute. Default is ``any()`` if no vertex weights nor
+            triangle weights are provided. Otherwise the default is ``sum()`` on
+            the first vertex weights column (or on the
+            first triangle weights column, if no vertex weights are provided).
+        replace : bool
+            Whether to replace the cached mesh. Setting this to False indicates
+            that the mesh did not change since the previous call, and should
+            lead to a performance boost. Defaults to True.
+        """
+        global CACHED_SOURCE
+        from .glyphs import Triangles
+        from .reductions import FloatingReduction, summary, sum as sum_rdn
+
+        cols = vertices.columns.tolist()
+        x, y, weights = cols[0], cols[1], cols[2:]
+
+        # Verify the simplex data structure
+        assert simplices.values.shape[1] >= 3, ('At least three vertex columns '
+                                                'are required for the triangle '
+                                                'definition')
+        simplices_all_ints = simplices.dtypes.iloc[:3].map(
+            lambda dt: np.issubdtype(dt, np.integer)
+        ).all()
+        assert simplices_all_ints, ('Simplices must be integral. You may '
+                                    'consider casting simplices to integers '
+                                    'with ".astype(int)"')
+
+        # Choose reduction based on vertex coordinate data
+        weight_col = None
+        if weights:
+            if agg is None:
+                agg = sum_rdn(weights[0])
+            weights_have_floats = vertices.dtypes[weights].map(
+                lambda dt: np.issubdtype(dt, np.float)
+            ).any()
+            assert not weights_have_floats or (
+                weights_have_floats and isinstance(agg, FloatingReduction)
+            ), ('Floating point vertex weights (zs) require aggs to be derived '
+                'from FloatingReduction')
+        elif agg is None:
+            assert simplices.values.shape[1] > 3, 'If no vertex weight column is provided, a triangle weight column is required.'
+            weight_col = simplices.columns[3]
+            agg = sum_rdn(weight_col)
+
+        if replace or CACHED_SOURCE is None:
+            if (isinstance(vertices, dd.DataFrame) and
+                    isinstance(simplices, dd.DataFrame)):
+                # TODO: Avoid .compute() calls, and add winding auto-detection
+                vals = vertices.values.compute()[simplices.values[:, :3].compute()]
+                vals = vals.reshape(np.prod(vals.shape[:2]), vals.shape[2])
+                CACHED_SOURCE = pd.DataFrame(vals, columns=vertices.columns)
+                if weight_col is not None:
+                    CACHED_SOURCE[weight_col] = simplices.values[:, 3].compute().repeat(3)
+            else:
+                simplex_coords = simplices.values[:, :3]
```

Comment:
Not sure if this is ready for review but in testing I found that if I add a weight column to the simplices, this array and the ``simplices.values[0, :3]`` array used below become float types causing float indexing errors.

Reply:
Thanks for letting me know. This issue should be fixed now.